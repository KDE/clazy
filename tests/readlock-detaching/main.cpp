#include <QtCore/QObject>
#include <QtCore/QString>
#include <QtCore/QMap>
#include <QtCore/QReadWriteLock>


class Test
{
    QReadWriteLock m_projectLock;
    QMap<QString, QString> m_fileToProjectParts;
    QMap<QString, QString> m_someOtherMap;

    void test(const QString &fileName)
    {
        m_someOtherMap.find(fileName); // OK, we did not pretend to lock anything here...
        QReadLocker locker(&m_projectLock);
        auto it = m_fileToProjectParts.find(fileName); // WARN, possible detach
        auto lookup = m_fileToProjectParts[fileName]; // WARN, possible detach
        Q_UNUSED(lookup);
        auto it3 = QMap<QString, QString>().find(fileName); // OK, not a member
    }

    void test2(const QString &fileName)
    {
        QWriteLocker locker(&m_projectLock);
        auto it = m_fileToProjectParts.find(fileName); // OK, we have a write lock
    }

    void test3(const QString &fileName)
    {
        {
            QReadLocker locker(&m_projectLock);
            auto it = m_fileToProjectParts.constFind(fileName); // OK - const version of this function
        }

        auto it = m_fileToProjectParts.find(fileName); // OK, outside of locker
    }

    void testUnlock(const QString &fileName)
    {
        QReadLocker locker(&m_projectLock);
        auto it = m_fileToProjectParts.find(fileName); // WARN
        locker.unlock();
        it = m_fileToProjectParts.find(fileName); // OK, we unlocked
    }

    void testLockUnlock(const QString &fileName)
    {
        auto it = m_fileToProjectParts.find(fileName); // OK, we did not lock yet
        m_projectLock.lockForRead();
        it = m_fileToProjectParts.find(fileName); // WARN, inside of read-only lock
        m_projectLock.unlock();
        it = m_fileToProjectParts.find(fileName); // OK, we unlocked
    }
};
