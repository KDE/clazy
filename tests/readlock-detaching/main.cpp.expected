readlock-detaching/main.cpp:17:19: warning: Possibly detaching a member while inside of a read-only mutex scope [-Wclazy-readlock-detaching]
readlock-detaching/main.cpp:18:23: warning: Possibly detaching a member while inside of a read-only mutex scope [-Wclazy-readlock-detaching]
readlock-detaching/main.cpp:42:19: warning: Possibly detaching a member while inside of a read-only mutex scope [-Wclazy-readlock-detaching]
readlock-detaching/main.cpp:51:14: warning: Possibly detaching a member while inside of a read-only mutex scope [-Wclazy-readlock-detaching]
